### ISSUE TITLE
Update language with url params resulted mixing page locale

### ISSUE THREAD

**Issue by anumang** [intershop/intershop-pwa#207]
**Actual Behavior**

Using `lang` parameter depending on configuration set-up causing mixed page locale selection.

**Expected Behavior**
 
 Given language by using `lang` parameter should reflect completely on current page locale without depending on your environment setup.

**Steps to Reproduce the Bug**
    Strange behaviour exist if follow steps to set-up environment and execute;
* Set-up **different initial locales** between;
  * `LOCALE_ID` which is used to set default locale at `configuration.module.ts:47` (Able to control explicitly by using `providers: [{provide: LOCALE_ID, useValue: 'fr-FR'}]`)
  * First available of locale from locales at `environment.prod.ts or environment.ts` (Current setup using `en_US` as a first available locale)
* Assume we have `fr_FR` as default locale from `LOCALE_ID` and `en_US` as first locale from `environment.prod.ts or environment.ts`
* Run application, `npm start`
* Try to navigate `http://localhost:4200/home;lang=fr_FR`
* See mixing over selected locale and page content.
  * ![image](https://user-images.githubusercontent.com/5325022/79960608-e2465b80-8485-11ea-8a14-cbe131a85f46.png)

**Environment Details**

*Desktop: *
 - Chrome 80.0.3987.163
 - PWA 0.18

**Possible Root Cause**
* Application causing repeating language  switch within short period of time, which triggers known `ngx-translate` issue ngx-translate/core#1030,
* As a result application state  switched correct language, loaded translation may differs,
* Adding  defer times while switching language like using `debounceTime` or  removing possible repeated language switch with same language code two times or more by using  `translateService.use or translateService.setDefaultLang` which is leading translation loads in wrong order.

**Comment by dhhyi** [comment]
Hi @anumang,

sorry for the late replay. I've been very busy working on #215. The new view renderer ivy will bring some amazing new features to the PWA 😁.

Using [LOCALE_ID](https://angular.io/api/core/LOCALE_ID) from Angular is something, that we do not support. And yes, our documentation is not quite clear on that; I'll work on that...

Currently we are basically only using ngx-translate for localization. However to use pipes on templates for formatting dates, numbers and prices, we have to build on top of the Angular localization framework. That's why we've built wrappers around the native pipes to instruct the Angular pipes on-the-fly with the current locale, that is managed in our NgRx state.

Up until Angular 9 the localization of Angular was always a build-time setting. This means if you would like to run the same code for different locales, you would have to deploy multiple instances of it. Lastly with our support for multiple channels, this was not an option for us and we decided to use ngx-translate. For the Angular part of our application the base locale should be ignored and setting it to different values has therefor no or even strange effects.

So I'm sorry to say, but this is a classic won't fix.

Some time after the Angular 9 upgrade we might think about dropping ngx-translate in favor for the native runtime support.

If I may ask, what is the original problem you would have liked to solve with this approach?

**Comment by anumang** [comment]
Hi @dhhyi , 

Seems my issue miss understood because issue is not related with usage or even existence of `LOCALE_ID`. In our code, we never updated or used `LOCALE_ID` other than existing usage on ISH PWA code at version 0.18. 

I just used it in the description to show how to re-produce the same issue with minimum change on your clean code at version 0.18 so that you can inspect further details. Just try to make it simple.

For our case, we have different browser locale setting (as `LOCALE_ID`) and different first placed locale (as first available locale or initial selected locale) at locales array in environment files. 

To create a combination of this issue, just we need to provide;
* `LOCALE_ID` or `window.navigator.locale` should have different value than given first locale in locales list
* Run application with express server,
* Try to navigate `http://localhost:4200/home;lang=${LOCALE_ID}` 

We applied temporary fix in our environment depending on ngx-translate/core#1030 issue which I believed the root cause of this issue. So I also mentioned at last section.

**Comment by dhhyi** [comment]
So, let me get this right: you want something like #104? (setting language based on browser settings)

This is something that's still in our backlog 😅.

**Comment by anumang** [comment]
:)
@dhhyi, the issue causing to show translations with wrong locale while application state has running with correct locale and ending up translations & product details are shown in two different locale. So it should not be evaluated duplication or part of a feature.

Just preventing possible race condition  between locale changes solve easily ngx-translate/core#1030 issue which is the root cause. 

So just by using `debounceTime` between locale updates we can easily prevent trouble ngx-translate/core#1030 issue;
https://github.com/intershop/intershop-pwa/blob/41d9d2ef2fa0a3945fca17de668215428f45874f/src/app/core/store/locale/locale.effects.ts#L23-L29

Instead of this, we can use with `debounceTime` with safety duration;
```js
  @Effect({ dispatch: false })
  setLocale$ = this.store.pipe(
    select(getCurrentLocale),
    mapToProperty('lang'),
    whenTruthy(),
    debounceTime(250),
    tap(lang => this.translateService.use(lang))
  );
```

**Comment by dhhyi** [comment]
Yes, NOW I get it. Sorry it took me so long, must be one of my bad days. 😅 

Debouncing actually works. Just re-scheduling with delay zero does also work for me. Can you please double-check? Thank you for pointing me the right way.

Do you mind if I add you as an author to that commit? #karma 😉

**Comment by anumang** [comment]
@dhhyi, thank you for understanding and resolving the issue :tada:. 

At the end, We just made a temporary solution but still I am happy to be part of it :) So it's not necessity but as you wish if you count me as an author for this fix.

My long term expectations,

Translation library  should  bring stable fix for their issue if they decide to wont-fix for their race condition issue and suggest developers to update locales only once, you may need to re-evaluate ISH locales architecture / logic from scratch or may switch another possible solutions which you mentioned with possible changes after angular 9 upgrade.
